### PR TITLE
Do continuous integration and codecoverage as in reduceSPI

### PR DESCRIPTION
**Pull request by geoffwoollard**
Some notes on what I did that deserve a closer look

1. Badges in README.md. What am I supposed to use for the links at the top of https://raw.githubusercontent.com/compSPI/reduceSPI/master/README.md

2. reduceSPI just has tests in `tests/test_*.py`. However ioSPI has  `tests/test_*.py` and also `tests/test_iotools/test_*.py`. What should be in `ioSPI/ci_codecov/pytest.ini` to handle the extra folder? https://github.com/geoffwoollard/ioSPI/commit/b668095c68ead853a60bde18259bb0025f83c9f1

3. I took the isort and black transformers out of `.deepsource.toml` https://github.com/geoffwoollard/ioSPI/commit/d99cc15204dd94ed4f6bfed62f8513d2f4adb893

4. `.codecov.yml` has an extra part to ignore folders that are absent in reduceSPI. I removed `coverage` folder because its empty in ioSPI and absent in reduceSPI. https://github.com/geoffwoollard/ioSPI/commit/1980c1ef66f468ebd2ea534884dbc1ab220f1e44

5. `.github/workflows/build.yml` is quite a bit different. Note that the flags for black are absent. https://github.com/geoffwoollard/ioSPI/commit/6311a5169e9229ce13fc56bde575d9eea0c04030#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721L51 These were important to have the linting and testing pass.

6. Aside: there's a folder `iospi/iotools` and also `ioSPI/iotools/`. They have different .py files in them. Is it supposed to be like that?

### CODE REVIEW

**Comment by geoffwoollard** [comment]
I'm struggling to satisfy isort/flake8/black at the same time. what is the workflow? I'm isorting, but then flake8 has an issue

```
(ioSPI) dhcp-128-189-219-129:ioSPI gw$ isort /Users/gw/repos/ioSPI/tests/test_fourier.py
(ioSPI) dhcp-128-189-219-129:ioSPI gw$ git status
On branch ci_codecov
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	modified:   tests/test_fourier.py

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   tests/test_fourier.py

(ioSPI) dhcp-128-189-219-129:ioSPI gw$ git commit tests/test_fourier.py -m 'isorted'
Check for byte-order marker..............................................Passed
Check for case conflicts.................................................Passed
Check for merge conflicts................................................Passed
Check Yaml...........................................(no files to check)Skipped
Mixed line ending........................................................Passed
Don't commit to branch...................................................Passed
Check for added large files..............................................Passed
Trim Trailing Whitespace.................................................Passed
black....................................................................Passed
isort....................................................................Passed
blacken-docs.............................................................Passed
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

tests/test_fourier.py:1:1: D100 Missing docstring in public module
tests/test_fourier.py:3:1: I202 Additional newline in a group of imports. 'from ioSPI import fourier' is identified as Third Party and 'import numpy' is identified as Third Party.

(ioSPI) dhcp-128-189-219-129:ioSPI gw$
```

**Comment by geoffwoollard** [comment]
I'll try this https://github.com/PyCQA/flake8-import-order/issues/145

**Comment by ninamiolane** [comment]
@geoffwoollard the pre-commit hook flake8 does not correct the files automatically, it is just failing. In this case, we need to write the error messages and address them. Here:
- there is a docstring missing at the very top of the test_fourier.py file (isort cannot write a docstring on its own)
- the second one is a bit weirder, it seems to say that you need an additional blank line before this ordeR? Did you try this?

> Badges in README.md. What am I supposed to use for the links at the top of https://raw.githubusercontent.com/compSPI/reduceSPI/master/README.md

Does it work to just replace reduceSPI by ioSPI?

> reduceSPI just has tests in tests/test_*.py. However ioSPI has tests/test_*.py and also tests/test_iotools/test_*.py. What should be in ioSPI/ci_codecov/pytest.ini to handle the extra folder? geoffwoollard@b668095

Let's use tests/test_*.py for now. We'll organize them in subfolders when we have more tests: bottom-up approach = we do it only when we need it.


> I took the isort and black transformers out of .deepsource.toml geoffwoollard@d99cc15

Yay thanks.


> .codecov.yml has an extra part to ignore folders that are absent in reduceSPI. I removed coverage folder because its empty in ioSPI and absent in reduceSPI. geoffwoollard@1980c1e

Cool. thanks.

> .github/workflows/build.yml is quite a bit different. Note that the flags for black are absent. geoffwoollard@6311a51#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721L51 These were important to have the linting and testing pass.

OK!

> Aside: there's a folder iospi/iotools and also ioSPI/iotools/. They have different .py files in them. Is it supposed to be like that?

No, it's not meant to be like this: ioSPI/iotools is redundant, since we know that we will have iotools in ioSPI. Let's use the bottom up approach again, and only have one `ioSPI` folder, with *.py modules in it. We'll subdivide this in subfolders when we need to (not before!).

**Comment by geoffwoollard** [comment]
The problem is that flake8 wants

```
import numpy as np
from ioSPI import fourier
```

And isort wants

```
import numpy as np

from ioSPI import fourier
```

When I commit with the hooks, isort changes things to 

```
import numpy as np

from ioSPI import fourier
```

Then flake8 fails.

This link says I need to tell flake8 somehow that there should be a line between them, because ioSPI is local, not third party.

**Comment by geoffwoollard** [comment]
Actually this works
```
import numpy as np

from ..ioSPI import fourier
```

**Comment by geoffwoollard** [comment]
There is an issue with iospi and ioSPI. When I checked out the files inside iospi from the main branch, then went inside ioSPI. So now on my branch there is no iospi, only ioSPI folder. Hopefully it stays that way when I merge and there is only only the ioSPI folder and no more iospi.
Screen Shot 2021-10-04 at 9 03 53 PM

File: .flake8 [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[flake8]
+docstring-convention = numpy
+import_order_style = smarkets
+max-line-length = 88
+extend-ignore = E203
```
Comment (ninamiolane) [inline .flake8:5]: @geoffwoollard how about adding `I202` to the extend-ignore line?

**Comment by geoffwoollard** [comment]
I was really confused for a while that the CI couldn't find this module. Locally it was in ioSPI, but actually git saw it as in iospi. So I just copied it (cp, not git mv) and renamed it as cryodataset.py. 2b7884268a53f1f1d824e432dd1525c43ceb766e

```
(ioSPI) dhcp-128-189-219-129:iotools gw$ git mv cryo_dataset.py cryodataset.py
fatal: not under version control, source=ioSPI/iotools/cryo_dataset.py, destination=ioSPI/iotools/cryodataset.py
```

**Comment by geoffwoollard** [comment]
@fredericpoitevin maybe you can write some tests for `atomic_models.py`?

```
----------- coverage: platform linux, python 3.9.5-final-0 -----------
Name                             Stmts   Miss  Cover   Missing
--------------------------------------------------------------
ioSPI/__init__.py                    1      0   100%
ioSPI/fourier.py                    67     17    75%   26-33, 54-61, 104
ioSPI/iotools/__init__.py            0      0   100%
ioSPI/iotools/atomic_models.py      51     35    31%   45, 47, 54, 76-80, 102-110, 127-134, 156-168
ioSPI/iotools/cryodataset.py        71     12    83%   36, 38-39, 47-50, 56-59, 87
--------------------------------------------------------------
TOTAL                              190     64    66%
Coverage XML written to file coverage.xml

FAIL Required test coverage of 90.0% not reached. Total coverage: 66.32%

============================= 15 passed in 12.03s ==============================
Error: Process completed with exit code 1.
```

File: .codecov.yml [hunk 2]
```diff
@@ -4,22 +4,22 @@ coverage:
   status:
     project:
       default:
-        target: 90%
+        target: 50%
```
Comment (ninamiolane) [inline .codecov.yml:7]: This should be 90% though: this is the threshold under which the PR will not pass: we don't want a PR to pass if we don't have a least 90% of the code that is covered, i.e. tested by the unit-tests. Why the decrease here?

File: .codecov.yml [hunk 3]
```diff
@@ -4,22 +4,22 @@ coverage:
...
-        if_not_found: success
+        if_not_found: error
         if_ci_failed: error
         informational: False
         only_pulls: false
     patch:
       default:
         # basic
-        target: 90%
+        target: 50%
```
Comment (ninamiolane) [inline .codecov.yml:22]: Ditto

File: .codecov.yml [hunk 4]
```diff
@@ -28,11 +28,11 @@ coverage:
...
-        if_not_found: success
+        if_not_found: error
         if_ci_failed: error
         only_pulls: false
 
 # Files to ignore
 ignore:
   - "data"
-  - "coverage"
+  - "notebooks"
```
Comment (ninamiolane) [inline .codecov.yml:38]: We will have to test the notebooks actually, but we can leave this for later I guess.

File: .flake8 [hunk 5]
```diff
@@ -0,0 +1,6 @@
+[flake8]
+docstring-convention = numpy
+import_order_style = smarkets
+max-line-length = 88
+extend-ignore = E203,I202,I100
```
Comment (ninamiolane) [inline .flake8:5]: Just curious, what is I100? Do you know why I didn't need these exclude in reduceSPI?

**Review by ninamiolane — COMMENTED** [review]
Looking good! We can merge this since most of it is working, and it will give us a chance to see if everything else is in place at the next PR.

But only if you can tackle the 90% code coverage threshold ASAP?

File: .flake8 (hunk 5 above)
Comment (geoffwoollard) [reply .flake8:5]: It's very similar as I202, with the spacing between the imports. Flake does not know that ioSPI is local, and wants to put the imports together, while isort knows and splits them up. See https://github.com/compSPI/ioSPI/pull/19/commits/2d7e03c2d619bcc9caabc177e0907bc2940eeb40

In your tests you do https://github.com/compSPI/reduceSPI/blob/master/tests/test_offline.py#L3, but there is no other import like `import numpy` or something nearby. My guess is that you would have the same issue.

**Comment by geoffwoollard** [comment]
I wrote some tests, but the CI is not happening? Where did they go?

**Comment by ninamiolane** [comment]
Oh yeah, I had the same problem with reduceSPI at some point... and then it came back! I don't know what is happening, TBH it could be that github action is experiencing a bug, or latency?

Maybe let's try to merge this, and see what it gives on a new PR?

**Comment by ninamiolane** [comment]
Oh lol, it's back x)

**Comment by geoffwoollard** [comment]
There's some problem with the code coverage. I wrote tests for parts that is says aren't covered. They are special because they are @jitted, so maybe there's some issue with that.

```
----------- coverage: platform linux, python 3.9.5-final-0 -----------
Name                             Stmts   Miss  Cover   Missing
--------------------------------------------------------------
ioSPI/__init__.py                    1      0   100%
ioSPI/fourier.py                    67     16    76%   26-33, 54-61
ioSPI/iotools/__init__.py            0      0   100%
ioSPI/iotools/atomic_models.py      51      1    98%   159
ioSPI/iotools/cryodataset.py        71     12    83%   36, 38-39, 47-50, 56-59, 87
--------------------------------------------------------------
TOTAL                              190     29    85%
```

The tests are https://github.com/geoffwoollard/ioSPI/blob/ci_codecov/tests/test_fourier.py#L51 and https://github.com/geoffwoollard/ioSPI/blob/ci_codecov/tests/test_fourier.py#L65

**Comment by ninamiolane** [comment]
Let's merge and see what codecov says. Thanks for this 🎉

**Review by ninamiolane — APPROVED** [review]

**Pull request merged**